### PR TITLE
Toolbar reactification

### PR DESCRIPTION
**Pull request by BeatC**
This PR is intended to move generation of primary and secondary toolbars into react application. Some UI logic related to toolbars are still in the "old app" and will be refactored in the second PR. The main goal of this one was to create `Toolbar` and `ToolbarButton` component that use logic from the old app.

### CODE REVIEW

**Comment by damencho** [comment]
Jenkins test this please.

File: react/features/base/styles/shimStyles.web.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Shim style properties to work correctly on Web.
```
Comment (saghul) [inline react/features/base/styles/shimStyles.web.js]: We haven't reached an agreement of how to do styles in React, so let's not go there yet.

File: react/features/room-lock/components/RoomLockPrompt.web.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { Component } from 'react';
```
Comment (saghul) [inline react/features/room-lock/components/RoomLockPrompt.web.js]: Please drop this since @damencho is working on dialogs on React. Also, why is this needed for toolbars?

File: react/features/toolbar/components/Toolbar.web.js [hunk 3]
```diff
@@ -0,0 +1,409 @@
+/* @flow */
+
+declare var AJS: Object;
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: please add the "global' comment too, that way we'll remember to get rid of them eventually.

File: react/features/toolbar/components/Toolbar.web.js [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+declare var APP: Object;
+declare var config: Object;
+declare var interfaceConfig: Object;
+declare var $: Function;
+
+import { connect } from 'react-redux';
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: alphabetical order

File: react/features/toolbar/components/Toolbar.web.js [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+     * Constructor of Toolbar component.
+     *
+     * @param {Object} props - The read-only React Component props with which
+     * the new instance is to be initialized.
+     */
+    constructor(props) {
+        super(props);
+
+        // Bind methods to save the context
+        const self: any = this;
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: unnecessary, you can just do

`this._foo = this._foo.bind(this);`

File: react/features/toolbar/components/Toolbar.web.js [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+                acc[place].set(button.id, button);
+            }
+
+            return acc;
+        }, {
+            primaryToolbar: new Map(),
+            secondaryToolbar: new Map()
+        });
+
+        this.setState({
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: declare the initial state

File: react/features/toolbar/components/Toolbar.web.js [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+     */
+    static propTypes = AbstractToolbar.propTypes;
+
+    /**
+     * Defines buttons to render in primary and secondary toolbars.
+     *
+     * @inheritdoc
+     * @returns {void}
+     */
+    componentWillMount() {
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: you can do all this in the construtor

File: react/features/toolbar/components/Toolbar.web.js [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+        this._registerListeners();
+
+        // Initialise the recording module.
+        if (config.enableRecording) {
+            const eventEmitter = APP.UI.eventEmitter;
+
+            Recording.init(eventEmitter, config.recordingType);
+        }
+
+        // Display notice message at the top of the toolbar
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: please add a TODO note so we can refactor this later

File: react/features/toolbar/components/Toolbar.web.js [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+
+        APP.UI.addListener(UIEvents.LOCAL_RAISE_HAND_CHANGED,
+            this._onLocalRaiseHandChanged);
+
+        APP.UI.addListener(UIEvents.FULLSCREEN_TOGGLED,
+            this._onFullScreenToggled);
+
+        APP.UI.addListener(UIEvents.SHOW_CUSTOM_TOOLBAR_BUTTON_POPUP,
+            this._onShowCustomToolbarPopup);
+
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: extra newlines?

File: react/features/toolbar/components/Toolbar.web.js [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+    }
+
+    /**
+     * Event handler for side toolbar container toggled event.
+     *
+     * @param {string} containerId - ID of the container.
+     * @returns {void}
+     */
+    _onSideToolbarContainerToggled(containerId) {
+        Object.keys(DEFAULT_TOOLBAR_BUTTONS).forEach(
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: use for..of

File: react/features/toolbar/components/Toolbar.web.js [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+     *
+     * @param {boolean} isRaisedHand - Flag showing whether hand is raised.
+     * @returns {void}
+     */
+    _onLocalRaiseHandChanged(isRaisedHand) {
+        const buttonId = 'toolbar_button_raisehand';
+        const button = this.state.secondaryToolbar.get(buttonId);
+
+        if (isRaisedHand) {
+            button.toggled = true;
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: `button.toggled = !!isRaisehand;`

File: react/features/toolbar/components/Toolbar.web.js [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+     *
+     * @param {Object} button - Object representing the button.
+     * @returns {ReactElement}
+     * @private
+     */
+    _renderToolbarButton(button) {
+        const onClick = event => {
+            const handler = BUTTON_HANDLERS[button.id];
+
+            if (!$(event.target).prop('disabled') && handler) {
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: do we really need jquery here?

File: react/features/toolbar/components/Toolbar.web.js [hunk 13]
```diff
@@ -0,0 +1,409 @@
...
+            if (button) {
+                const place = this._getToolbarButtonPlace(buttonName);
+
+                button.buttonName = buttonName;
+                acc[place].set(button.id, button);
+            }
+
+            return acc;
+        }, {
+            primaryToolbar: new Map(),
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: I think arrays are a better representation here, where each element is an object with stuff

File: react/features/toolbar/components/Toolbar.web.js [hunk 14]
```diff
@@ -0,0 +1,409 @@
...
+     *
+     * @param {string} popupSelectorID - The selector id of the popup to show.
+     * @param {boolean} show - True or false/show or hide the popup.
+     * @param {number} timeout - The time to show the popup.
+     * @returns {void}
+     */
+    _onShowCustomToolbarPopup(popupSelectorID, show, timeout) {
+        const gravity = $(popupSelectorID).attr('tooltip-gravity');
+
+        AJS.$(popupSelectorID)
```
Comment (saghul) [inline react/features/toolbar/components/Toolbar.web.js]: can't we get rid off AJS and use something from atlaskit here?

File: react/features/toolbar/components/ToolbarButton.web.js [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+     * @param {Object} props - The read-only properties with which the new
+     * instance is to be initialized.
+     */
+    constructor(props: Object) {
+        super(props);
+
+        this.popups = [];
+
+        // Bind methods to save the context
+        const self: any = this;
```
Comment (saghul) [inline react/features/toolbar/components/ToolbarButton.web.js]: unnecessary

File: react/features/toolbar/components/ToolbarButton.web.js [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    /**
+     * Adds popup to the array property.
+     *
+     * @param {HTMLElement} popup - Popup element.
+     * @returns {void}
+     * @private
+     */
+    _pushPopup(popup) {
```
Comment (saghul) [inline react/features/toolbar/components/ToolbarButton.web.js]: do we really need a helper function for pushing to an array?

File: react/features/toolbar/components/ToolbarButton.web.js [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+
+/**
+ * Represents a button in Toolbar on React.
+ *
+ * @class ToolbarButton
+ * @extends AbstractToolbarButton
+ */
+export default class ToolbarButton extends AbstractToolbarButton {
+
+    /**
```
Comment (saghul) [inline react/features/toolbar/components/ToolbarButton.web.js]: declare proptypes

File: react/features/toolbar/components/secondaryToolbarHandlers.js [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+                   name="sipNumber" type="text"
+                   value="${defaultNumber}" autofocus>`;
+
+    APP.UI.messageHandler.openTwoButtonDialog({
+        titleKey,
+        msgString,
+        leftButtonKey: 'dialog.Dial',
+
+        // eslint-disable-next-line max-params
+        submitFunction: (e, v, m, f) => {
```
Comment (saghul) [inline react/features/toolbar/components/secondaryToolbarHandlers.js]: please use descriptive variable names

File: react/features/toolbar/components/ToolbarButton.web.js [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+            }
+
+            return (
+                <div
+                    className = { popup.className }
+                    data-i18n = { popup.dataAttr }
+                    data-tooltip = { gravity }
+                    id = { popup.id }
+                    key = { popup.id }
+                    ref = { this._pushPopup } />
```
Comment (saghul) [inline react/features/toolbar/components/ToolbarButton.web.js]: I think this is a very obscure way to populate the popups array.

File: react/features/toolbar/components/ToolbarButton.web.js (hunk 16 above)
Comment (BeatC) [inline react/features/toolbar/components/ToolbarButton.web.js]: Actually we don't but eslint requires not to add arrow functions to props. But it seems like we can pass `this.popups.push` to ref and get rid of the helper function.

File: react/features/toolbar/components/Toolbar.web.js (hunk 3 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: Ok

File: react/features/room-lock/components/RoomLockPrompt.web.js (hunk 2 above)
Comment (BeatC) [inline react/features/room-lock/components/RoomLockPrompt.web.js]: Because `AbstractToolbar` depends on that module.

File: react/features/base/styles/shimStyles.web.js (hunk 1 above)
Comment (BeatC) [inline react/features/base/styles/shimStyles.web.js]: I put the code for web platform from the old repository because `AbstractToolbar` depends on styles module.

File: react/features/toolbar/components/Toolbar.web.js (hunk 5 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: I can't because Flow doesn't allow to do this.

File: react/features/toolbar/components/Toolbar.web.js (hunk 7 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: Thank you, but I moved toolbar buttons to toolbar store.

File: react/features/toolbar/components/Toolbar.web.js (hunk 12 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: Intention of this PR was just to start generate toolbar and toolbar buttons via React. I've got rid of jquery in the code based on this branch.

File: react/features/toolbar/components/Toolbar.web.js (hunk 13 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: Arrays suits here but in some cases we need to get elements by key. If we represent data via arrays it costs O(n) to access the element. Maps allow us to get cheap element access (O(1) mostly) and we can traverse them in loop (because they are ordered) just like arrays.

File: react/features/toolbar/components/Toolbar.web.js (hunk 8 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: Created separate component for notice message in new branch.

File: react/features/toolbar/components/Toolbar.web.js (hunk 14 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: We should but seems like it out of scope of this PR.

File: react/features/toolbar/components/Toolbar.web.js (hunk 10 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: Did in further updates.

File: react/features/toolbar/components/Toolbar.web.js (hunk 11 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: did in further updates.

File: react/features/toolbar/components/Toolbar.web.js (hunk 5 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: But we can remove self if explicitly declare property type.

File: react/features/toolbar/components/Toolbar.web.js (hunk 3 above)
Comment (BeatC) [inline react/features/toolbar/components/Toolbar.web.js]: Flow and Eslint contradict each other.

File: react/features/toolbar/components/ToolbarButton.web.js (hunk 19 above)
Comment (BeatC) [inline react/features/toolbar/components/ToolbarButton.web.js]: we use popups array only for translations. Since I've added translation of toolbar button after every update we don't need this array at all.

**Comment by saghul** [comment]
Work continues in #1397, which already includes this one but goes further in making it fully-react.
